rfcs/0000-terraform-workflow-improvements.md

```diff
@@ -0,0 +1,826 @@
+
+- Testing modules requires two module calls – one to create IAM resources and one to create other resources
+- Whoever is running tests (user or automation) will still require IAM permissions in order to test modules
+
+#### Dev Deployment
+
+- Dev deployment requires prerequisite deployment of IAM resource module
+- Engineers do not require IAM permissions to deploy changes to dev
+
+#### Prod Deployment
+
+- Prod deployment requires prerequisite deployment of IAM resource module
+- Engineers do not require IAM permissions to deploy changes to prod
+
+#### Risks
+
+- Onerous process to deploy IAM modules separately from application modules
+- Module testing requires creating IAM resources separately
+- Engineers can still define IAM resources anywhere, though they'll lack the permissions to apply them
+
+# Prior art
+[prior-art]: #prior-art
+
+<!--
+Discuss prior art, both the good and the bad, in relation to this proposal. A few examples of what
+this can include are:
+
+- Can we learn something about this proposal from other projects in CTD?
+- Can we learn something about this proposal from other transit agencies?
+- Can we learn something about this proposal from past experiences in other jobs or projects?
+
+This section is intended to encourage you as an author to think about the lessons from other places,
+and provide readers of your RFC with a fuller picture.
+
+If there is no prior art, that is fine.
+
+Note that while precedent is some motivation, it does not on its own motivate an RFC.
+-->
+
+Workflow automation for Terraform is entirely new territory for the authors of this document. We've done a fair amount of research and factfinding to formulate a plan that we think brings CTD's usage of Terraform more in line with industry best practices, and takes advantage of products that take some of the work off our shoulders. That said, it's entirely possible there's something we missed, and we're open to recommendations of solutions, software or patterns from anyone who cares to weigh in.
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+<!--
+- What parts of the design do you expect to resolve through the RFC process before this gets merged?
+- What parts of the design do you expect to resolve through the implementation of this feature
+  before stabilization?
+- What related issues do you consider out of scope for this RFC that could be addressed in the
+  future independently of the solution that comes out of this RFC?
+-->
+
+One major unresolved question is the timeline for implementing all the improvements laid out in this RFC. As noted in the [Drawbacks][drawbacks] section above, it's a lot of work. Once this RFC is approved we'll need to scope out the work and divide it accordingly between the infrastructure and engineering teams.
+
+Other questions for the reader:
+
+- Is the proposed [root module naming convention][root-module-naming-convention] sensible and sufficiently comprehensive?
```

Comment:
It could use an example, real-world or fake.

Reply:
Hm, does the [Proposed Module Names](https://github.com/mbta/technology-docs/blob/idw-terraform-workflow-improvements/rfcs/0000-terraform-workflow-improvements.md#proposed-module-names) section not satisfy this need?

Reply:
I meant an example of what one of those modules would look like. What's contained within one root module's directory. One application right now has multiple files across the `devops/terraform/` directory tree. I may be conflating "root module" and "directory" because we currently group things that way.

My motivation for asking this was that for the sorts of terraform changes I do regularly, I want a better idea of what my experience making a change to my app would be like in this naming / organization scheme.

Reply:
I think your day-to-day experience won't change at all, because you'd be working _within_ a module that has a set name per the recommendations in this document. How you structure individual files within your root modules (and even child modules) is up to your team's discretion, though Terraform [does offer recommendations](https://developer.hashicorp.com/terraform/language/modules/develop/structure) for standard filenames such as `main.tf`, `variables.tf`, and `outputs.tf`.

Let me know if that answers your question.

Reply:
Okay, I'm understanding it a bit more. I do think an example would still help make it clearer, and could prevent a little bit of "am I putting this in the right place" when implementing this. As an application developer, I don't think of it as root modules and child modules, I think of it as which files am I editing, so a concrete example would help me summarize and digest all of the complicated parts of this RFC.

Let's say there's a team with a relatively standard structure with 1 or 2 apps that each have prod/staging/dev-green environments, some global resources that aren't per environment (like s3 buckets), and references the shared ctd-app-ecs. What repos/files does the team care about in the new organization, which files are referenced with versions, and which files need tests?

Reply:
I think what you're calling out is important knowledge to have when using Terraform, but on a more fundamental level than this RFC aims to cover. I do think that anyone writing Terraform configuration needs to understand the concept of child modules and root modules, and the relationship between the two. Terraform [offers a tutorial](https://developer.hashicorp.com/terraform/tutorials/modules) that covers these concepts.

I added a note about the assumptions this document makes about knowledge of Terraform's concepts: https://github.com/mbta/technology-docs/pull/16/commits/37e3ddb77b797bd0abab685b2f34a0859a08d316